### PR TITLE
Add example deployment for Kubernetes

### PR DESCRIPTION
**Pull request by meibensteiner**

### CODE REVIEW

**Review by dyladan — APPROVED** [review]

File: example/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+
```
Comment (arminru) [inline example/README.md:1]:
```suggestion
```

File: example/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+# FluentD to Dynatrace Log Ingest
```
Comment (arminru) [inline example/README.md:2]: Please link to the example folder from the top-level README so it's easier to discover.

File: example/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+It is possible to have FluentD ingest the logs of your kubernetes nodes and let FluentD send the resulting logs to your Dynatrace environment. See below for instructions.
+## Prerequisites
+Generally needed
```
Comment (arminru) [inline example/README.md:8]:
```suggestion
Generally needed:
```

File: example/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+- A Dynatrace namespace, if you do not already have one
+- An [API Token](https://www.dynatrace.com/support/help/dynatrace-api/basics/dynatrace-api-authentication/) with "Ingest logs" permission
+- The dynatrace-monitoring service-account
+
+   ``kubectl apply -f service-account.yaml``
+
+Only needed in combination with activegate
```
Comment (arminru) [inline example/README.md:16]:
```suggestion
Only needed in combination with a Dynatrace ActiveGate:
```
That's the proper brand name.

File: example/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
+
+# FluentD to Dynatrace Log Ingest
+
+It is possible to have FluentD ingest the logs of your kubernetes nodes and let FluentD send the resulting logs to your Dynatrace environment. See below for instructions.
+## Prerequisites
```
Comment (arminru) [inline example/README.md]: Please format the file using markdownlint or similar.

File: example/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+Generally needed
+- A Dynatrace namespace, if you do not already have one
+- An [API Token](https://www.dynatrace.com/support/help/dynatrace-api/basics/dynatrace-api-authentication/) with "Ingest logs" permission
+- The dynatrace-monitoring service-account
+
+   ``kubectl apply -f service-account.yaml``
+
+Only needed in combination with activegate
+- A [PaaS Token](https://www.dynatrace.com/support/help/get-started/access-tokens/)
+- Available persistent storage in your cluster
```
Comment (arminru) [inline example/README.md:19]:
```suggestion
- Available persistent storage in your Kubernetes cluster
```
To distinguish it from the DT cluster.

File: example/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+## Decide on a log ingest
```
Comment (arminru) [inline example/README.md:21]: This heading sounds unfinished... on a log ingest endpoint/method/...?

File: example/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+You have two options when choosing the endpoint fluentd sends your log to.
+This is either an activegate, which might already exist in your environment or a prexisting log ingest on your SaaS Dynatrace environment.
+### Option 1: Deploy an ActiveGate
+
+1. Create a secret holding the environment URL and login credentials for this registry, making sure to replace.
```
Comment (arminru) [inline example/README.md:28]:
```suggestion
1. Create a secret holding the environment URL and login credentials for this registry, making sure to replace
```

File: example/README.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+   * ${YOUR_ENVIRONMENT_URL} with your environment URL (without 'http'). Example: environment.live.dynatrace.com
```
Comment (arminru) [inline example/README.md]:
```suggestion
   * ${YOUR_ENVIRONMENT_URL} with your environment URL without 'https' (e.g., `env123456.live.dynatrace.com`)
```

File: example/README.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+   * ${YOUR_ENVIRONMENT_ID} with the ID in your environment URL
```
Comment (arminru) [inline example/README.md]:
```suggestion
   * ${YOUR_ENVIRONMENT_ID} with the ID of your environment (e.g., `env123456`)
```

File: example/README.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+   * ${YOUR_PAAS_TOKEN} with the PaaS token you created in Prerequisites
```
Comment (arminru) [inline example/README.md]:
```suggestion
   * ${YOUR_PAAS_TOKEN} with the PaaS token you created as described in the Prerequisites
```

File: example/README.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+   ``kubectl create secret docker-registry tenant-docker-registry --docker-server=${YOUR_ENVIRONMENT_URL} --docker-username=${YOUR_ENVIRONMENT_ID} --docker-password=${YOUR_PAAS_TOKEN} -n dynatrace``
+
+2. Edit the file named configmap-activegate.yaml, making sure to replace
+   * ${YOUR_ENVIRONMENT_ID} with the ID in your environment URL.
```
Comment (arminru) [inline example/README.md]:
```suggestion
   * ${YOUR_ENVIRONMENT_ID} with the ID of your environment (e.g., `env123456`)
```

File: example/README.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+   * ${YOUR_CLUSTER_ID} with your cluster id.
```
Comment (arminru) [inline example/README.md]: Please add an example here as well

File: example/README.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    
+     ``kubectl get namespace kube-system  -o jsonpath='{.metadata.uid}'``
+   
+   * fluent.conf with an configuration suiting your environment. Our example is using the kubernetes-metadata-filter to enrich ingested log lines with information about its underlying kubernetes infrastructure. **Depending on the applications on your cluster you might need to manually modify the fluent.conf.**
```
Comment (arminru) [inline example/README.md]:
```suggestion
   * your `fluent.conf` file with a configuration suiting your environment. Our example is using the kubernetes-metadata-filter to enrich ingested log lines with information about its underlying kubernetes infrastructure. **Depending on the applications on your cluster you might need to manually modify the `fluent.conf` file.**
```

File: example/README.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+3. Edit the file named activegate.yaml, making sure to replace.
```
Comment (arminru) [inline example/README.md:44]:
```suggestion
3. Edit the file named `activegate.yaml`, making sure to replace
```

File: example/README.md [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+   * ${YOUR_ENVIRONMENT_URL} with your environment URL (without 'http'). Example: environment.live.dynatrace.com
```
Comment (arminru) [inline example/README.md]:
```suggestion
   * ${YOUR_ENVIRONMENT_URL} with your environment URL without 'https' (e.g., `env123456.live.dynatrace.com`)
```

File: example/README.md [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+4. That same activegate.yaml also deploys a persistent volume to queue log entries that were not pushed to Dynatrace yet. Make sure you have a persistent volume or storage class readily available.
```
Comment (arminru) [inline example/README.md:47]:
```suggestion
4. That same `activegate.yaml` also deploys a persistent volume to queue log entries that were not pushed to Dynatrace yet. Make sure you have a persistent volume or storage class readily available.
```

File: example/README.md [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+5. Deploy Configmap and ActiveGate
+
+   ``kubectl apply -f activegate.yaml configmap-activegate.yaml``
+
+### Option 2: Use the log ingest of your managed SaaS Dynatrace environment
+
+Not only your activegate presents an endpoint, but also your Dynatrace environment is able to ingest logs.
```
Comment (arminru) [inline example/README.md:55]:
```suggestion
Not only your ActiveGate presents an endpoint, but also your Dynatrace environment is able to ingest logs directly.
```

File: example/README.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+1. Edit the file named configmap-saas.yaml, making sure to replace
```
Comment (arminru) [inline example/README.md:57]:
```suggestion
1. Edit the file named `configmap-saas.yaml`, making sure to replace
```

File: example/README.md [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+   ``kubectl apply -f service-account.yaml``
+
+Only needed in combination with activegate
+- A [PaaS Token](https://www.dynatrace.com/support/help/get-started/access-tokens/)
+- Available persistent storage in your cluster
+
+## Decide on a log ingest
+
+You have two options when choosing the endpoint fluentd sends your log to.
+This is either an activegate, which might already exist in your environment or a prexisting log ingest on your SaaS Dynatrace environment.
```
Comment (arminru) [inline example/README.md:24]:
```suggestion
This is either an ActiveGate, which might already exist in your environment or a pre-existing log ingest on your SaaS Dynatrace environment.
```

File: example/README.md [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+
+1. Create a secret holding the environment URL and login credentials for this registry, making sure to replace.
+
+   * ${YOUR_ENVIRONMENT_URL} with your environment URL (without 'http'). Example: environment.live.dynatrace.com
+   * ${YOUR_ENVIRONMENT_ID} with the ID in your environment URL
+   * ${YOUR_PAAS_TOKEN} with the PaaS token you created in Prerequisites
+
+   ``kubectl create secret docker-registry tenant-docker-registry --docker-server=${YOUR_ENVIRONMENT_URL} --docker-username=${YOUR_ENVIRONMENT_ID} --docker-password=${YOUR_PAAS_TOKEN} -n dynatrace``
+
+2. Edit the file named configmap-activegate.yaml, making sure to replace
```
Comment (arminru) [inline example/README.md:36]:
```suggestion
2. Edit the file named `configmap-activegate.yaml`, making sure to replace
```

File: example/README.md [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+   ``kubectl apply -f activegate.yaml configmap-activegate.yaml``
+
+### Option 2: Use the log ingest of your managed SaaS Dynatrace environment
+
+Not only your activegate presents an endpoint, but also your Dynatrace environment is able to ingest logs.
+
+1. Edit the file named configmap-saas.yaml, making sure to replace
+   * ${YOUR_ENVIRONMENT_ID} with the ID in your environment URL.
+   * ${YOUR_CLUSTER_ID} with your cluster id.
```
Comment (arminru) [inline example/README.md]: Same format suggestions as above.

File: example/README.md [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+    
+     ``kubectl get namespace kube-system  -o jsonpath='{.metadata.uid}'``
+   
+   * fluent.conf with an configuration suiting your environment. Our example is using the kubernetes-metadata-filter to enrich ingested log lines with information about its underlying kubernetes infrastructure. **Depending on the applications on your cluster you might need to manually modify the fluent.conf.**
```
Comment (arminru) [inline example/README.md]: Same format suggestions as above.

File: example/README.md [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+
+   ``kubectl apply -f configmap-saas.yaml``
+
+## Build FluentD docker image
+
+Build the FluentD docker image provided in our example and upload it to your repository. **Depending on the applications running on your cluster you might need to add plugins.**
+
+## Deploy FluentD 
+
+1. Create a secret with your API Token. Make sure to replace ${YOUR_API_TOKEN} with the API Token that has log ingest permissions.
```
Comment (arminru) [inline example/README.md:75]:
```suggestion
1. Create a secret with your Dynatrace API Token. Make sure to replace ${YOUR_API_TOKEN} with the API Token that has log ingest permissions.
```

File: example/README.md [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+
+   ``kubectl create secret generic tokens --from-literal="log-ingest=${YOUR_API_TOKEN}" -n dynatrace``
+
+2. Replace ${YOUR_FLUENTD_IMAGE} with the fluentd image you built
```
Comment (arminru) [inline example/README.md]:
```suggestion
2. Replace `${YOUR_FLUENTD_IMAGE}` in `fluentd.yaml` with the fluentd image you built 
```

File: example/README.md [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+
+## Deploy FluentD 
+
+1. Create a secret with your API Token. Make sure to replace ${YOUR_API_TOKEN} with the API Token that has log ingest permissions.
+
+   ``kubectl create secret generic tokens --from-literal="log-ingest=${YOUR_API_TOKEN}" -n dynatrace``
+
+2. Replace ${YOUR_FLUENTD_IMAGE} with the fluentd image you built 
+
+3. Deploy fluentd.yaml
```
Comment (arminru) [inline example/README.md:81]:
```suggestion
3. Deploy `fluentd.yaml`
```

File: README.md [hunk 27]
```diff
@@ -82,6 +82,10 @@ It is recommended to leave this optional configuration set to `false` unless abs
 
 If set to true, the timestamp of the log message will be injected into the body of the message in a format readable by Dynatrace.
 
+## Example
+
+An example of the necessary components used for Kubernetes is documented [here](https://github.com/dynatrace-oss/fluent-plugin-dynatrace/example).
```
Comment (arminru) [inline README.md:87]:
```suggestion
## Example setup in Kubernetes

An full example demonstrating how to set up Fluentd on Kubernetes and export logs to Dynatrace can be found in the [example folder](./example).
```
Maybe?

File: example/README.md [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+You have two options when choosing the endpoint fluentd sends your log to.
+This is either an ActiveGate, which might already exist in your environment or a pre-existing log ingest on your SaaS Dynatrace environment.
+
+### Option 1: Deploy an ActiveGate
+
+1. Create a secret holding the environment URL and login credentials for this registry, making sure to replace
+
+   - ${YOUR_ENVIRONMENT_URL} with your environment URL (without 'http'). Example: environment.live.dynatrace.com
+   - ${YOUR_ENVIRONMENT_ID} with the ID in your environment URL
+   - ${YOUR_PAAS_TOKEN} with the PaaS token you created in Prerequisites
```
Comment (arminru) [inline example/README.md:31]: You could put all of these placeholders in backticks as well (`${YOUR_ENVIRONMENT_URL}`, ...) for consistent formatting (same below).

**Review by arminru — APPROVED** [review]

File: example/README.md [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+
+   - `${YOUR_ENVIRONMENT_URL}` with your environment URL without 'https' (e.g., `env123456.live.dynatrace.com`)
+   - `${YOUR_ENVIRONMENT_ID}` with the ID of your environment (e.g., `env123456`)
+   - `${YOUR_PAAS_TOKEN}` with the PaaS token you created as described in the Prerequisites
+
+   ``kubectl create secret docker-registry tenant-docker-registry --docker-server=${YOUR_ENVIRONMENT_URL} --docker-username=${YOUR_ENVIRONMENT_ID} --docker-password=${YOUR_PAAS_TOKEN} -n dynatrace``
+
+2. Edit the file named `configmap-activegate.yaml`, making sure to replace
+   - `${YOUR_ENVIRONMENT_ID}` with the ID of your environment (e.g., `env123456`)
+   - `${YOUR_CLUSTER_ID}` with your cluster id.
```
Comment (arminru) [inline example/README.md:37]:
```suggestion
   - `${YOUR_CLUSTER_ID}` with your Kubernetes Cluster ID which can be retrieved using this command:
```

File: example/README.md [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+
+   ``kubectl apply -f activegate.yaml configmap-activegate.yaml``
+
+### Option 2: Use the log ingest of your managed SaaS Dynatrace environment
+
+Not only your ActiveGate presents an endpoint, but also your Dynatrace environment is able to ingest logs directly.
+
+1. Edit the file named `configmap-saas.yaml`, making sure to replace
+   - `${YOUR_ENVIRONMENT_ID}` with the ID of your environment (e.g., `env123456`)
+   - `${YOUR_CLUSTER_ID}` with your cluster id.
```
Comment (arminru) [inline example/README.md:58]:
```suggestion
   - `${YOUR_CLUSTER_ID}` with your Kubernetes Cluster ID which can be retrieved using this command:
```

**Review by arminru — APPROVED** [review]

File: example/README.md [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+1. Create a secret holding the environment URL and login credentials for this registry, making sure to replace
+
+   - `${YOUR_ENVIRONMENT_URL}` with your environment URL without 'https' (e.g., `env123456.live.dynatrace.com`)
+   - `${YOUR_ENVIRONMENT_ID}` with the ID of your environment (e.g., `env123456`)
+   - `${YOUR_PAAS_TOKEN}` with the PaaS token you created as described in the Prerequisites
+
+   ``kubectl create secret docker-registry tenant-docker-registry --docker-server=${YOUR_ENVIRONMENT_URL} --docker-username=${YOUR_ENVIRONMENT_ID} --docker-password=${YOUR_PAAS_TOKEN} -n dynatrace``
+
+2. Edit the file named `configmap-activegate.yaml`, making sure to replace
+   - `${YOUR_ENVIRONMENT_ID}` with the ID of your environment (e.g., `env123456`)
```
Comment (arminru) [inline example/README.md:36]:
```suggestion
   - `${YOUR_ENVIRONMENT_URL}` with your Dynatrace environment URL without 'https' (e.g., `env123456.live.dynatrace.com`)
   - `${YOUR_ENVIRONMENT_ID}` with the ID of your Dynatrace environment (e.g., `env123456`)
   - `${YOUR_PAAS_TOKEN}` with the PaaS token you created as described in the Prerequisites

   ``kubectl create secret docker-registry tenant-docker-registry --docker-server=${YOUR_ENVIRONMENT_URL} --docker-username=${YOUR_ENVIRONMENT_ID} --docker-password=${YOUR_PAAS_TOKEN} -n dynatrace``

2. Edit the file named `configmap-activegate.yaml`, making sure to replace
   - `${YOUR_ENVIRONMENT_ID}` with the ID of your Dynatrace environment (e.g., `env123456`)
```

File: example/README.md [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+   - `${YOUR_CLUSTER_ID}` with your Kubernetes Cluster ID which can be retrieved using this command:
+
+     ``kubectl get namespace kube-system  -o jsonpath='{.metadata.uid}'``
+
+   - your `fluent.conf` file with a configuration suiting your environment. Our example is using the kubernetes-metadata-filter to enrich ingested log lines with information about its underlying kubernetes infrastructure. **Depending on the applications on your cluster you might need to manually modify the `fluent.conf` file.**
+
+3. Edit the file named `activegate.yaml`, making sure to replace
+   - `${YOUR_ENVIRONMENT_URL}` with your environment URL (without 'https' (e.g., `env123456.live.dynatrace.com`)
```
Comment (arminru) [inline example/README.md:44]:
```suggestion
   - `${YOUR_ENVIRONMENT_URL}` with your Dynatrace environment URL (without 'https' (e.g., `env123456.live.dynatrace.com`)
```

File: example/README.md [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+5. Deploy Configmap and ActiveGate
+
+   ``kubectl apply -f activegate.yaml configmap-activegate.yaml``
+
+### Option 2: Use the log ingest of your managed SaaS Dynatrace environment
+
+Not only your ActiveGate presents an endpoint, but also your Dynatrace environment is able to ingest logs directly.
+
+1. Edit the file named `configmap-saas.yaml`, making sure to replace
+   - `${YOUR_ENVIRONMENT_ID}` with the ID of your environment (e.g., `env123456`)
```
Comment (arminru) [inline example/README.md:57]:
```suggestion
   - `${YOUR_ENVIRONMENT_ID}` with the ID of your Dynatrace environment (e.g., `env123456`)
```

**Review by arminru — APPROVED** [review]

**Comment by arminru** [comment]
Thanks, @meibensteiner!

**Pull request merged**